core/src/main/java/hudson/model/Computer.java

```diff
@@ -939,7 +946,12 @@ public String getSearchUrl() {
      * If this is the master, it returns the system property of the master computer.
      */
     public EnvVars getEnvironment() throws IOException, InterruptedException {
-        return EnvVars.getRemote(getChannel());
+        if (cachedEnvironment != null) {
+            return cachedEnvironment;
```

Comment:
indeed, need to rely on a second volatile "environmentCached" flag (like cachedHostName does).